src/_app/render.tsx

```diff
@@ -14,12 +14,18 @@ export default function renderApp({
 }) {
   const { useApp } = options;
   const customAppFile = getCustomAppFile({ options });
-  let AppComponent;
+  let AppComponent = DefaultApp;
 
-  if (useApp && customAppFile?.client?.default) {
-    AppComponent = customAppFile.client.default;
-  } else {
-    AppComponent = DefaultApp;
+  if (useApp) {
+    if (options.useDocument) {
+      if (customAppFile?.server.default) {
+        AppComponent = customAppFile.server.default;
+      }
+    } else {
+      if (customAppFile?.client.default) {
+        AppComponent = customAppFile.client.default;
+      }
+    }
```

Comment:
Are you sure we're using the proper `customAppFile` here? It seems to me that the `customAppFile` returned to the user at initial render should always be the one initialised in `client` environment (no matter what `useDocument` value).

This bring us back to the current -I'd say- architectural limitation of `next-page-tester`: we currently never perform a full tree render in `server` mode (like Next.js obviously does during SSR).

We directly (and roughly) return an element tree similar to the one generated in-browser after React hydration. This solution works fine as long as we consider `_app` and pages.

When it comes to deal with Next.js `_document` (and `next/head`), this approach shows its limitations. This is because Next.js SSR renders the whole tree on the server (providing `next/head` with a special context to update head elements during SSR phase), and then mounts the client application with React.hydrate on `#__next` div.

We're currently leaving out this complexity and we might have to find out whether we can work around this limitations or we should provide one or two different ways of rendering with `next-page-tester`.

Out of topic: we might be technically able to return from `getPage` with:
- a `html` string representing the SSR-rendered page to inject in JSDOM
- an elements' tree to mount with (React.hydrate) on the `#__next` div

This would definitely be closer to how I understand Next.js works (ad also a promising approach), but we should find out how to get a good dev experience out of it.

Reply:
I agree, we should be returning the `client` customApp file to the user. Problem is that the client version fails at the `renderToStaticMarkup` in this specific use-case.

I think its a compromise at this point:
1. Have `_app` correctly render server side (I think this means client side events wont work in `_app` if you had any -- e.g `onClick` handlers or `Link` -- I have never used this and would expect is not so frequent)
1. Rendering `_app` with wont work for some projects and will crash

As you mentioned, we are already hacking a lot of stuff and there will surely be limitations/tradeoffs with the current implementation.

With this (and other) PRs I just wanted to focus on more realistic examples of NextJS applications so we have those covered as much as possible. While everything is perfect for simple "todo apps", I'm afraid it still doesn't work for a lot of non-trivial projects out there.